geode-core/src/main/java/org/apache/geode/distributed/internal/ServerLocator.java

```diff
@@ -378,24 +379,40 @@ public void endResponse(Object request, long startTime) {
   }
 
 
-
-  private List<ServerLocation> getLocators() {
-    if (cachedLocators != null) {
+  private List<ServerLocation> getLocators(boolean requestInternal) {
+    if (cachedLocators != null && cachedRequestInternalLocators == requestInternal) {
       return cachedLocators;
     } else {
       synchronized (cachedLocatorsLock) {
         List<ControllerProfile> profiles = advisor.fetchControllers();
         List<ServerLocation> result = new ArrayList<>(profiles.size() + 1);
         for (ControllerProfile profile : profiles) {
-          result.add(buildServerLocation(profile));
+          result.add(buildServerLocation(profile, requestInternal));
         }
-        result.add(new ServerLocation(hostNameForClients, port));
+        String host;
+        if (requestInternal) {
+          host = hostName;
+        } else {
+          host = hostNameForClients;
+        }
+        result.add(new ServerLocation(host, port));
+        cachedRequestInternalLocators = requestInternal;
         cachedLocators = result;
         return result;
       }
     }
   }
 
+  protected static ServerLocation buildServerLocation(GridProfile p, boolean requestInternal) {
+    String host;
+    if (requestInternal) {
+      host = p.getInternalHost();
+    } else {
+      host = p.getHost();
+    }
+    return new ServerLocation(host, p.getPort());
+  }
+
   protected static ServerLocation buildServerLocation(GridProfile p) {
```

Comment:
Could uses of this method be replaced with calls to the new, two-argument method instead, to reduce redundancy?